src/VBORenderer.cc

```diff
@@ -0,0 +1,467 @@
+ *  follow the requirements of the GNU GPL in regard to all of the
+ *  software in the executable aside from CGAL.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ */
+
+#include "VBORenderer.h"
+#include "polyset.h"
+#include "csgnode.h"
+#include "printutils.h"
+
+#include <cstddef>
+
+VBORenderer::VBORenderer()
+	: Renderer()
+{
+    const char *vs_source =
+    "uniform float xscale, yscale;\n"
+    "attribute vec4 color1, color2;\n"
+    "attribute vec3 pos_b, pos_c;\n"
+    "attribute vec3 trig, mask;\n"
+    "varying vec4 fcolor1, fcolor2;\n"
+    "varying vec3 tp, tr;\n"
+    "varying float shading;\n"
+    "void main() {\n"
+    "  vec4 p0 = gl_ModelViewProjectionMatrix * gl_Vertex;\n"
+    "  vec4 p1 = gl_ModelViewProjectionMatrix * vec4(pos_b, 1.0);\n"
+    "  vec4 p2 = gl_ModelViewProjectionMatrix * vec4(pos_c, 1.0);\n"
+    "  float a = distance(vec2(xscale*p1.x/p1.w, yscale*p1.y/p1.w), vec2(xscale*p2.x/p2.w, yscale*p2.y/p2.w));\n"
+    "  float b = distance(vec2(xscale*p0.x/p0.w, yscale*p0.y/p0.w), vec2(xscale*p1.x/p1.w, yscale*p1.y/p1.w));\n"
+    "  float c = distance(vec2(xscale*p0.x/p0.w, yscale*p0.y/p0.w), vec2(xscale*p2.x/p2.w, yscale*p2.y/p2.w));\n"
+    "  float s = (a + b + c) / 2.0;\n"
+    "  float A = sqrt(s*(s-a)*(s-b)*(s-c));\n"
+    "  float ha = 2.0*A/a;\n"
+    "  gl_Position = p0;\n"
+    "  tp = mask * ha;\n"
+    "  tr = trig;\n"
+    "  vec3 normal, lightDir;\n"
+    "  normal = normalize(gl_NormalMatrix * gl_Normal);\n"
+    "  lightDir = normalize(vec3(gl_LightSource[0].position));\n"
+    "  shading = 0.2 + abs(dot(normal, lightDir));\n"
+    "  fcolor1 = color1;\n"
+    "  fcolor2 = color2;\n"
+    "}\n";
+
+    /*
+    Inputs:
+    tp && tr - if any components of tp < tr, use color2 (edge color)
+    shading  - multiplied by color1. color2 is is without lighting
+    */
+    const char *fs_source =
+    "varying vec4 fcolor1, fcolor2;\n"
+    "varying vec3 tp, tr, tmp;\n"
+    "varying float shading;\n"
+    "void main() {\n"
+    "  gl_FragColor = vec4(fcolor1.r * shading, fcolor1.g * shading, fcolor1.b * shading, fcolor1.a);\n"
+    "  if (tp.x < tr.x || tp.y < tr.y || tp.z < tr.z)\n"
+    "    gl_FragColor = fcolor2;\n"
+    "}\n";
+
+		GLint status;
+		GLenum err;
+    GLuint vs = glCreateShader(GL_VERTEX_SHADER);
+    glShaderSource(vs, 1, (const GLchar**)&vs_source, NULL);
+    glCompileShader(vs);
+		err = glGetError();
+    if (err != GL_NO_ERROR) {
+        PRINTDB("OpenGL Error: %s\n", gluErrorString(err));
```

Comment:
should stop compiling the shader (`return` or throw an exception to terminate) if there are problems here - this will lead to error log spam